src/liballoc/collections/btree/navigate.rs

```diff
@@ -123,36 +123,30 @@ impl<'a, K, V> Handle<NodeRef<marker::Immut<'a>, K, V, marker::Leaf>, marker::Ed
 impl<'a, K, V> Handle<NodeRef<marker::Mut<'a>, K, V, marker::Leaf>, marker::Edge> {
     /// Moves the leaf edge handle to the next leaf edge and returns references to the
     /// key and value in between.
-    /// Unsafe for two reasons:
-    /// - The caller must ensure that the leaf edge is not the last one in the tree.
-    /// - Using the updated handle may well invalidate the returned references.
+    /// Unsafe because the caller must ensure that the leaf edge is not the last one in the tree.
     pub unsafe fn next_unchecked(&mut self) -> (&'a mut K, &'a mut V) {
         unsafe {
             let kv = replace(self, |leaf_edge| {
                 let kv = leaf_edge.next_kv();
                 let kv = unwrap_unchecked(kv.ok());
                 (ptr::read(&kv).next_leaf_edge(), kv)
             });
-            // Doing the descend (and perhaps another move) invalidates the references
-            // returned by `into_kv_mut`, so we have to do this last.
+            // Doing this last is faster, according to benchmarks.
             kv.into_kv_mut()
         }
     }
 
     /// Moves the leaf edge handle to the previous leaf and returns references to the
     /// key and value in between.
-    /// Unsafe for two reasons:
-    /// - The caller must ensure that the leaf edge is not the first one in the tree.
-    /// - Using the updated handle may well invalidate the returned references.
```

Comment:
This does return references, so should the comment stay or is there some other reason for it to be wrong?

Reply:
Well, this PR implies it is no longer a problem? The references stay valid while using the handle (and, I assume, until dropping it).

Reply:
What does "using the updated handle" even mean? Use for what? Calling `next` again certainly must work, indeed. But are there other uses, ones that to not work, such as `push` or so?

Reply:
I don't know. I think it's possible that all of the public and even private methods in the auxiliary modules would operate fine given updated handles and respect the invariants of node.rs (but, as always they, would very likely violate the much stricter invariants of map.rs). But they are extremely likely to trample on the previously returned references.

So I guess it should be a postcondition on the caller: only use the updated handle for next or next_back? But I feel like a lot of the auxiliary functions could be describing how they're supposed to be used, and that description will always turn out to be wrong later.

Reply:
This "postcondition" instead becomes statically type information in the upcoming commit.

Reply:
I think all I am saying is, it seems better to keep this comment for functions that actually return (pairs of) references.

Reply:
I don't understand why. I claim "Using the updated handle may well invalidate the returned references." is no longer possible, unless you invoke unsafe methods violating their preconditions, or worse unsafety. Unless you mean, "dropping the updated handle, will invalidate the returned references".

Reply:
I wrote that comment before seeing the `ValMut` matches. If the API now prevents the bad mutations, that's even better. :)